sources/api/migration/migrations/v1.8.0/cluster-dns-ip-list/src/main.rs

```diff
@@ -0,0 +1,204 @@
+#![deny(rust_2018_idioms)]
+
+use migration_helpers::{migrate, Migration, MigrationData, Result};
+use std::process;
+
+const CLUSTER_DNS_IP_KEY: &str = "settings.kubernetes.cluster-dns-ip";
+
+/// We changed `settings.kubernetes.cluster-dns-ip` to support being either a string or a list of strings.
+fn run() -> Result<()> {
+    migrate(ClusterDNSIPListMigration)
+}
+
+struct ClusterDNSIPListMigration;
+
+impl Migration for ClusterDNSIPListMigration {
+    /// New versions allow the older string values to be present, so we don't need to do anything.
+    fn forward(&mut self, input: MigrationData) -> Result<MigrationData> {
+        println!("ClusterDNSIPListMigration has no work to do on upgrade.");
+        Ok(input)
+    }
+
+    /// Older versions don't know about list-style settings, so we need to create a scalar setting using the first value.
+    fn backward(&mut self, mut input: MigrationData) -> Result<MigrationData> {
+        let maybe_prior_value = input.data.get(CLUSTER_DNS_IP_KEY);
+
+        // If the current value is a string, don't touch it.
+        if let Some(prior_value) = maybe_prior_value {
+            if prior_value.is_string() {
+                println!(
+                    "{} is already a string value ('{}'), and does not require migration.",
+                    CLUSTER_DNS_IP_KEY, prior_value
+                );
+                return Ok(input);
+            }
+        }
+
+        // If the current value is an array and the first element is a string, that element becomes the new value.
+        // Any other cases result in clearing the value.
+        let new_value = maybe_prior_value
+            .and_then(|dns_ip_value| {
+                println!(
+                    "Found existing value for '{}': '{}'",
+                    CLUSTER_DNS_IP_KEY, dns_ip_value
+                );
+                dns_ip_value.as_array()
+            })
+            .and_then(|ip_array| ip_array.iter().next())
+            .map(|ip_value| ip_value.clone());
```

Comment:
Checking my understanding here also -

If we have a value and it's the empty list, we'll have a `None` here, and remove the setting from the datastore. The previous template had a guard for whether the setting was defined, so this is safe - we'll render a config file without that section.

Reply:
Yes, that's right.